### ISSUE TITLE
Error 111 connecting to localhost:6379. Connection refused.

### ISSUE THREAD

**Issue by lironmo** [areebbeigh/codeinterview-backend#5]
Hi,

cool project!

I am trying to run the project locally and got :
Error 111 connecting to localhost:6379. Connection refused.

seems that the Redis is missing from the project - can you please give a full explanation of how to run locally? (I also add some values to the database in order to find the language )

**Comment by lironmo** [comment]
if it simply needs Redis (without ant configs) - no need to add -

but worth mention at the readme

**Comment by areebbeigh** [comment]
Hey, thanks. Yeah, you need redis (or any message queue) running.

You can quickly set one up with docker:

```
docker run --name my-redis -d -p 6379:6379 redis redis-server --appendonly yes
```

**Comment by lironmo** [comment]
Thank @areebbeigh.
I added also the following confs:

CELERY_RESULT_BACKEND = "redis"
CELERY_REDIS_HOST = "localhost"
CELERY_REDIS_PORT = 6379
CELERY_REDIS_DB = 0

what should I add to:
CELERY_BROKER_URL = 'pyamqp://'


it's seems that the sync doesn't works correctly (probably wrong setup)

**Comment by areebbeigh** [comment]
Broker url would be "redis://"

**Comment by lironmo** [comment]
still doesn't works 

configs 
`import os

from celery.schedules import crontab

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

SECRET_KEY = 'j%4s9n79np!^nrq3&h4=6a8r2c^ex9s)gg3s(zsx((o@qll2yj'
DEBUG = True
ALLOWED_HOSTS = []

# Celery config

CELERY_BROKER_URL = 'redis://'


CELERY_RESULT_BACKEND = "redis"
CELERY_REDIS_HOST = "localhost"
CELERY_REDIS_PORT = 6379
CELERY_REDIS_DB = 0

CELERY_TASK_ROUTES = {
    # WARNING: room.tasks still need explicit queue name when chaining.
    # see rooms.signals.dispatch_run_task. Help?
    'rooms.tasks.*': {'queue': 'callbacks'},
    'tasks.sandbox.run_user_code': {'queue': 'sandbox'}
}
# Comment out to disable auto-deletion
CELERY_BEAT_SCHEDULE = {
    'delete_rooms': {
        'task': 'rooms.tasks.delete_rooms',
        'schedule': crontab(minute='*/1'),
        'options': {'queue' : 'celery_beat'},
    },
}

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'corsheaders',
    'rest_framework',
    'django_filters',
    'channels',
    'rooms',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'code_interview.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            os.path.join(BASE_DIR, 'templates'),
        ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'code_interview.wsgi.application'
ASGI_APPLICATION = "code_interview.routing.application"

##### Channels-specific settings

redis_url = 'redis://'

# Channel layer definitions
# http://channels.readthedocs.io/en/latest/topics/channel_layers.html
CHANNEL_LAYERS = {
    # "default": {
    #     # This example app uses the Redis channel layer implementation channels_redis
    #     "BACKEND": "channels_redis.core.RedisChannelLayer",
    #     "CONFIG": {
    #         "hosts": [redis_url,],
    #     },
    # },
    "default": {
        "BACKEND": "channels.layers.InMemoryChannelLayer"
    }
}

CORS_ORIGIN_WHITELIST = [
    'http://localhost:8080' # front-end
]

# REST framework
default_renderers = []
if DEBUG:
    default_renderers = ['rest_framework.renderers.BrowsableAPIRenderer']

REST_FRAMEWORK = {
    'DEFAULT_RENDERER_CLASSES': [
        'rest_framework.renderers.JSONRenderer',
    ] + default_renderers,
}

# Database
# https://docs.djangoproject.com/en/2.2/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}


# Password validation
# https://docs.djangoproject.com/en/2.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/

STATIC_URL = '/static/'
`

branch master.

the logs are sync - but the code is not synced and also the run doesn't works

**Comment by lironmo** [comment]
@areebbeigh - I think that I am doing something wrong:
I run the following components:
backend + frontend + Redis

with the default configuration (python manga runserver / npm start / redis docker)

**Comment by areebbeigh** [comment]
> @areebbeigh - I think that I am doing something wrong:
> I run the following components:
> backend + frontend + Redis
> 
> with the default configuration (python manga runserver / npm start / redis docker)

What isn't working? Is the web app up? Run requests get created in the backend?

You also need to run the codeinterview-sandbox for actual code execution to happen.

**Comment by lironmo** [comment]
Hi @areebbeigh,
web page  - UP
I did not start the sandbox - did not know it's related.

so I need to build the docker image at https://github.com/areebbeigh/codeinterview-sandbox and how to run it? (celery -A run worker -l info??)

**Comment by lironmo** [comment]
ok, now I have a different issue with the sendbox  - does the send box need rabbitMQ?

Cannot connect to amqp://guest:**@127.0.0.1:5672//:

**Comment by lironmo** [comment]
i changed the configuration to redis - the sandbox is up/

# CELERY_BROKER_URL = os.environ.get('REDIS_URL')
# CELERY_RESULT_BACKEND = os.environ.get('REDIS_URL')

CELERY_BROKER_URL = 'redis://'


CELERY_RESULT_BACKEND = "redis"
CELERY_REDIS_HOST = "localhost"
CELERY_REDIS_PORT = 6379
CELERY_REDIS_DB = 0

**Comment by areebbeigh** [comment]
> ok, now I have a different issue with the sendbox  - does the send box need rabbitMQ?
> 
> Cannot connect to amqp://guest:**@127.0.0.1:5672//:

Yup, you have to point the sandbox workers to the same message queue the web app pushes to. You can use redis for this. amqp:// is rabbitmq

**Comment by lironmo** [comment]
I did it!
all seems to work fine but still, i have the same  problem:
1. sync is not working 
2. result is missing 

the web page is up and running.

maybe I need to connect the backend to the sandbox in some way?

**Comment by lironmo** [comment]
maybe something is missing in the database?

because I had to add manually the rows for the language.

**Comment by lironmo** [comment]
it seems that the request sent from the webpage to the backend and dispatch celery task via dispatch_run_task.

but it's seems that the celery sandbox did not act on the task dispatch - checking why now

**Comment by lironmo** [comment]
I think I found the problem - wsl issue...:

 the code that checks if the docker image exists returns a different exception and gets unhandled  -- I will build manually and fix it later.

the exception happens in   self.image = self.client.images.get(
                settings.DOCKER_IMAGE_FULL_NAME)
                
                getting :
              requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

**Comment by lironmo** [comment]
more issues  - 
docker runtime issue at wsl - remove secure runtime.

now it seems to be stack at the celery when I run it manually.
